crates/bevy_scene/src/scene_filter.rs

```diff
@@ -0,0 +1,104 @@
+use bevy_utils::HashSet;
+use std::any::{Any, TypeId};
+
+/// A filter used to control which types can be added to a [`DynamicScene`].
+///
+/// This scene filter _can_ be used more generically to represent a filter for any given type;
+/// however, note that its intended usage with `DynamicScene` only considers [components].
+/// Adding types that are not components will have no effect when used with `DynamicScene`.
+///
+/// [`DynamicScene`]: crate::DynamicScene
+/// [components]: bevy_ecs::prelude::Component
+pub enum SceneFilter {
+    /// Contains the set of permitted types by their [`TypeId`].
+    ///
+    /// Types not contained within this set should not be allowed to be saved to an associated [`DynamicScene`].
+    ///
+    /// [`DynamicScene`]: crate::DynamicScene
+    Allowlist(HashSet<TypeId>),
+    /// Contains the set of prohibited types by their [`TypeId`].
+    ///
+    /// Types contained within this set should not be allowed to be saved to an associated [`DynamicScene`].
+    ///
+    /// [`DynamicScene`]: crate::DynamicScene
+    Denylist(HashSet<TypeId>),
+}
+
+impl SceneFilter {
+    /// Create a new [allowlist](Self::Allowlist).
+    pub fn new_allowlist() -> Self {
+        Self::Allowlist(HashSet::new())
+    }
+
+    /// Create a new [denylist](Self::Denylist).
+    pub fn new_denylist() -> Self {
+        Self::Denylist(HashSet::new())
+    }
+
+    /// Allow the given type, `T`.
+    ///
+    /// If this filter is already set as a [denylist](Self::Denylist),
+    /// then it will be completely replaced by a new [allowlist](Self::Allowlist).
+    pub fn allow<T: Any>(self) -> Self {
+        self.allow_by_id(TypeId::of::<T>())
+    }
+
+    /// Allow the given type.
+    ///
+    /// If this filter is already set as a [denylist](Self::Denylist),
+    /// then it will be completely replaced by a new [allowlist](Self::Allowlist).
+    pub fn allow_by_id(mut self, type_id: TypeId) -> Self {
+        match self {
+            Self::Allowlist(ref mut list) => {
+                list.insert(type_id);
+                self
+            }
+            Self::Denylist(_) => Self::Allowlist(HashSet::from([type_id])),
```

Comment:
i think calling `allow` on a `SceneFilter::Denylist` should remove it from the list if present and else be a no-op. (and vice versa)

to me the semantics of "allow" suggest that the component should be included in the final scene, unless explicitly later overridden by a `deny` call. (and vice versa)

```rust
// under the current behaviour:
filter
    .allow::<A>() // Allowlist[A]
    .allow::<B>() // Allowlist[A, B]
    .deny::<B>() // Denylist[B]
    .allow::<C>(); // Allowlist[C]
// oh no! we never denied A but it's no longer allowed.

// suggested behaviour
filter
    .allow::<A>() // Allowlist[A]
    .allow::<B>() // Allowlist[A, B]
    .deny::<B>() // Allowlist[A]
    .allow::<C>() // Allowlist[A, C]
```

though this locks the list to being either an allowlist or denylist from initialisation, leaving the open question of what to do calling `allow`/`deny` on a `None` scene filter. 

perhaps `SceneFilter::None` could keep a set of allows as well as a set of denies and force the user to specify (e.g. `DynamicSceneBuilder::set_allowing`) and otherwise panic on calling `build`.

Reply:
I feel like swapping to a `None` variant should be like a reset. It would feel weird if it still contained the list, right? If users want to persist that list they can do so with a `mem::swap` and just take the list (and pass it to another filter if desired).